crates/cli-util/src/context.rs

```diff
@@ -0,0 +1,158 @@
+        let os_env = OsEnv::default();
+        Self::with_env(&os_env, opts)
+    }
+
+    /// Loading CliEnv with a custom OsEnv. OsEnv can be customised in cfg(test)
+    pub fn with_env(os_env: &OsEnv, opts: CommonOpts) -> Self {
+        // Load .env file. Best effort.
+        let maybe_dotenv = dotenv();
+
+        // color setup
+        // NO_COLOR=1 with any value other than "0" means user doesn't want colors.
+        // e.g.
+        //  NO_COLOR=1 (no colors)
+        //  NO_COLOR=true (no colors)
+        //  NO_COLOR=something (no colors)
+        //  NO_COLOR=0 or unset (yes *color* if term supports it)
+        let should_color = os_env
+            .get("NO_COLOR")
+            .map(|x| x == "0")
+            .unwrap_or_else(|| true);
+
+        // dumb terminal? no colors or fancy stuff
+        let smart_term = os_env
+            .get("TERM")
+            .map(|x| x != "dumb")
+            .unwrap_or_else(|| true);
+
+        // CLICOLOR_FORCE is set? enforce coloring..
+        // Se http://bixense.com/clicolors/ for details.
+        let force_colorful = os_env
+            .get("CLICOLOR_FORCE")
+            .map(|x| x != "0")
+            .unwrap_or_else(|| false);
+
+        let colorful = if force_colorful {
+            // CLICOLOR_FORCE is set, we enforce coloring
+            true
+        } else {
+            // We colorize only if it's a smart terminal (not TERM=dumb, nor pipe)
+            // and NO_COLOR is anything but "0"
+            let is_terminal = std::io::stdout().is_terminal();
+            is_terminal && smart_term && should_color
+        };
+
+        // Ensure we follows our colorful setting in our console utilities
+        // without passing the environment around.
+        dialoguer::console::set_colors_enabled(colorful);
+
+        // Setup logging from env and from -v .. -vvvv
+        tracing_subscriber::fmt()
+            .with_writer(std::io::stderr)
+            .with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
+            .with_max_level(opts.verbose.log_level_filter().as_trace())
+            .with_ansi(colorful)
+            .init();
+
+        // We only log after we've initialized the logger with the desired log
+        // level.
+        let loaded_dotenv = maybe_dotenv.ok();
+        match &loaded_dotenv {
+            Some(path) => {
+                info!("Loaded .env file from: {}", path.display())
+            }
+            None => info!("Didn't load '.env' file"),
+        };
+
+        Self {
+            confirm_mode: opts.confirm,
+            ui: opts.ui,
+            network: opts.network,
+            colors_enabled: colorful,
+            loaded_dotenv,
+        }
+    }
+
+    pub fn get() -> arc_swap::Guard<Arc<CliContext>> {
+        GLOBAL_CLI_CONTEXT.get_or_init(Default::default).load()
+    }
+
+    /// Sets the global context to the given value. In general, this should be called once on CLI
+    /// initialization.
```

Comment:
Could this be enforced by not using an `ArcSwap` in the `GLOBAL_CLI_CONTEXT` and instead directly storing the `CliContext`? Or do we need the `ArcSwap` for testing scenarios where one wants to update the `CliContext`?

Reply:
That's the intention, although we don't actually have tests that exercise that now.